Introduce issue forms and update debug output for forms

georglauterbach opened the pull request:
# Description

This PR introduces [GitHub forms](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository#creating-issue-forms) and slightly alters the debug output for better visibility.

The debug output changed in so far as now a self-written function prints the environment in a useful way, and we do not use `printenv` anymore, because I think `printenv` clutters the whole log with information no one really needs. I did this because with the new form, we encourage people to use `DMS_DEBUG=1` more, so I figured prettier log output / a non-cluttered log to begin with would be nice.

See [this maintainers discussion](https://github.com/orgs/docker-mailserver/teams/maintainers/discussions/34)

## Type of change

<!-- Delete options that are not relevant. -->

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

williamdes commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,155 @@
+  - type: input
+    id: docker-compose-version
+    attributes:
+      label: Docker version
```
```suggestion
      label: Docker compose version
```

georglauterbach replied:
I think Compose if a proper noun: <https://docs.docker.com/compose/>

georglauterbach commented:
I cancelled the first two workflows manually to not waste CPU time :)

wernerfred commented:
I will review carefully later today

casperklein commented on target/scripts/helper-functions.sh:
```diff
@@ -15,6 +15,20 @@ function escape
+function print_environment
+{
+  [[ ${DMS_DEBUG} == 0 ]] && return
+  echo
```
Is the empty line (single `echo`) on purpose? Asking, because we never used that when printing to stdout afaik.

georglauterbach replied:
This was muscle memory, trying to get a blank line here. We could get rid of this, no big deal, just formatting.

casperklein replied:
Yeah. I think the "BEGIN/END" marker are enough to make it readable/clear.

georglauterbach replied:
Done. [de84818e8389c57fe4ee19929a8a8f7949cbb9c6]

casperklein commented on target/scripts/start-mailserver.sh:
```diff
@@ -256,9 +256,8 @@ function _defunc
-_notify 'inf' 'ENVIRONMENT'
-[[ ${DMS_DEBUG} -eq 1 ]] && printenv
 
+print_environment
```
From a logic point of view, `print_environment` should "print environment". The decision if that function output something should not be made within that function, but like it was directly in `start-mailserver.sh`.

georglauterbach replied:
I see you'r point. I will revert this.

georglauterbach replied:
Done. [de84818e8389c57fe4ee19929a8a8f7949cbb9c6]

williamdes commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,169 @@
+  - type: dropdown
+    id: main-memory
+    attributes:
+      label: How much RAM is available?
```
CPU count could also be important

georglauterbach replied:
Good idea! I will add this later.

georglauterbach replied:
Done. [e556f64932f6ce98016cac6763f218a4ff01f97c]

wernerfred commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+      label: Code of conduct
+      description: By submitting this issue, you agree to follow our [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md).
+      options:
+        - label: I agree to follow this project's Code of Conduct
+          required: true
+        - label: I have read this project's Code of Conduct
+          required: true
```
Imho this is unnecessarily doubled. We do not want to close a deal with someone

georglauterbach replied:
I see where you're coming from. And I guess adding a checkbox won't help the situation much.

georglauterbach replied:
Done. [ba9ea5a85f812e811782bb27864d87ed2916c056]

wernerfred commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+          required: true
+        - label: I checked that all ports are open and not blocked by my ISP / hoster.
+          required: true
+        - label: Just to make sure you're not just clicking checkboxes, please do not check this checkbox.
```
laughed hard on this one

georglauterbach replied:
I was smiling too when I wrote it...

wernerfred commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+    attributes:
+      label: Issue guidelines and documentation
+      description: We as maintainers help others in our free time to solve issues submitted in this issue tracker. As a consequence, we demand that the documentation and README is read carefully and that the issue tracker was searched for prior issues.
+      options:
+        - label: I read the README, the documentation and searched the issue tracker and could not find such an issue
+          required: true
+        - label: I checked that all ports are open and not blocked by my ISP / hoster.
+          required: true
+        - label: Just to make sure you're not just clicking checkboxes, please do not check this checkbox.
+          required: false
```
too much imho. Last sentence as a friendly reminder should be enough

georglauterbach replied:
I'm going to remove the first checkbox, but I'd like to keep the others here. They comfort me as a maintainer, especially the last one.

georglauterbach replied:
Done. [ba9ea5a85f812e811782bb27864d87ed2916c056]

wernerfred commented on target/scripts/helper-functions.sh:
```diff
@@ -15,6 +15,18 @@ function escape
   echo "${1//./\\.}"
 }
 
+function print_environment
+{
+  _notify 'inf' 'BEGIN ENVIRONMENT'
+
+  for VARIABLE in "${!VARS[@]}"
+  do
+    echo "${VARIABLE}=${VARS[${VARIABLE}]}"
+  done
+
+  _notify 'inf' 'END ENVIRONMENT\n'
+}
+
```
Do we have any chance to mask sensitive values by default?

georglauterbach replied:
That would be a nice addition, but would require a lot of additional work on our side that is not worth it IMHO.

wernerfred replied:
Then maybe add an hint to remove sensitive values in this output :)

georglauterbach replied:
Done. [ba9ea5a85f812e811782bb27864d87ed2916c056]

wernerfred reviewed (dismissed):
Thanks for this inital block of work. All in all this seems fine. But imho the wording is a bit to "agressive". I know that we have to deal a lot with people who don't use the templates or don't provide enough information **but** this form is here to solve this problem :) 

As we do not want to kind of "scare" a few people from filing issue reports i vote for a friendlier wording here and there.

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+  - type: dropdown
+    id: operating-system
+    attributes:
+      label: What operating system is DMS running on?
```
No, but I want to know if someone is using DMS on Windows to save myself the time of finding it out myself. More information on this results in less time wasted on my / our end.

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+        - <= 1GB
+        - <= 2GB
+        - <= 4GB
+        - <= 8GB
```
I can adjust that, yeah.

georglauterbach replied:
Done. [ba9ea5a85f812e811782bb27864d87ed2916c056]

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+    attributes:
+      label:  When does this occur and what happened?
+      description: Tell us what happened. Use Markdown formatting for code!
+      placeholder: Although `DMS_DEBUG=1` is set, I see no debug.
```
A good idea. I will tend to it.

georglauterbach replied:
Done. [ba9ea5a85f812e811782bb27864d87ed2916c056]

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+    validations:
+      required: true
+  - type: dropdown
+    id: isa
```
ISA stands for Instruction Set Architecture:)

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+    attributes:
+      label: Docker version
+      description: Please show us the output of `docker --version` on the host your mail server is running on if you are using Docker. Note that if you are not using Docker, there is no official support provided by maintainers as DMS mainly focuses on Docker (and Docker Compose).
+      placeholder: Docker version 20.10.7, build f0df350
```
I would rather not, TBH 🙈 It's reasonable to assume that this is an example, made clear by this text being gray, not black like the rest of the text.

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+    attributes:
+      label: Docker compose version
+      description: Please show us the output of `docker-compose --version` on the host your mail server is running on. Note that if you are not using Docker Compose, there is no official support provided by maintainers as DMS mainly focuses on Docker COmpose (and Docker).
+      placeholder: docker-compose version 1.28.6, build 5db8d86f
```
Same as above.

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+    id: uname-a
+    attributes:
+      label: The output of `uname -a`
+      description: To double-check that the information above is correct, please paste the output of `uname -a` here.
```
That actually was kind of the idea, but I'm fine with the proposed changes.

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,189 @@
+description: File a bug report
+title: "[BUG] "
+labels:
+  - kind/possible bug
```
Fine with me. I will remove the label again.

georglauterbach replied:
Done. [ba9ea5a85f812e811782bb27864d87ed2916c056]

polarathene commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,203 @@
+    attributes:
+      label: What level of experience do you have with Docker and mail servers?
+      description: You are not obliged to answer this question. But answering it helps us better understand where possible causes of errors lie. Less experienced users are more likely do make "obvious" mistakes, which is fine, and by answering this question, we know to look for them as well. We encourage everyone to participate, regardless of their level of experience.
+      options:
+        - I'm very much a beginner
+        - I'm a novice
+        - I know containerization but have less experience with mail
+        - I have experience with mail but less experience with containerization
+        - I know both pretty well and I'm experienced
+        - I'm an expert
+        - Trust me, I'm a (computer) engineer
```
Little rephrasing, if you're ok with being a little less formal. The expert option seemed a bit redundant so I folded it into the last one.

```suggestion
      options:
        - I have no idea what I'm doing! (beginner)
        - I know the basics, I've used Docker before (novice)
        - I'm comfortable with Docker / containers, but I'm relatively new to all this mail stuff
        - I have experience with mail servers, but less experience working with containers
        - I know both pretty well actually (experienced)
        - Trust me, I'm a (computer) engineer! (expert)
```

georglauterbach replied:
I like it :D

polarathene commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,203 @@
+    id: terms-miscellaneous
+    attributes:
+      label: Miscellaneous
+      description: Please read these carefully and check the appropriate boxes.
```
Not required, but at a later point if these turn out to be common issues that get explained to users during troubleshooting, perhaps have both the ports and SSL troubleshooting as doc pages and link to them here.

May be relevant early on so they're not cursing by the time they get to the end :sweat_smile:

georglauterbach replied:
I moved this section to the beginning.

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,203 @@
+      description: Please paste the output of `uname -a` here.
+      placeholder: Linux swift 5.11.0-31-generic \#33-Ubuntu SMP Wed Aug 11 13:19:04 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
+    validations:
+      required: true
```
Good point. I chose a solution that should fit everyone.

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,203 @@
+    id: experience
+    attributes:
+      label: What level of experience do you have with Docker and mail servers?
+      description: You are not obliged to answer this question. But answering it helps us better understand where possible causes of errors lie. Less experienced users are more likely do make "obvious" mistakes, which is fine, and by answering this question, we know to look for them as well. We encourage everyone to participate, regardless of their level of experience.
```
👍🏼

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,203 @@
+    id: relevant-log-output
+    attributes:
+      label: Relevant log output
+      description: Show us relevant log output here. You can enable debug output by setting `DMS_DEBUG` to `1`.
```
👍🏼

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,203 @@
+    id: main-memory
+    attributes:
+      label: How much RAM is available?
+      description: Please tell us how much main memory is available for DMS.
```
Good point too. I again tried to solve this in a way that makes the reader aware of what to choose now. You should have a second look:)

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,203 @@
+      options:
+        - x86_64 / AMD64
+        - ARM v7
+        - ARM64
```
Fair point :+1

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,203 @@
+    id: cpu-count
+    attributes:
+      label: How many CPU cores are available?
+      description: Please tell us how many CPU cores are available for DMS. A vCore is a CPU core used in virtual machines, a physical core is a real CPU core.
```
I reworked this section. One now has to also answer the question of what system (physical host, VDS, VPS) is being used.

williamdes commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,209 @@
+name: Bug Report
```
```suggestion
name: Bug report
```

williamdes commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,209 @@
+      options:
+        - x86_64 / AMD64
+        - ARM64 / ARM v8 (and newer)
+        - ARM v7
```
What about ppc64el and others?
Maybe just add an "other" option

georglauterbach replied:
There is no image built to support this platform AFAIK, and we do not provide support for forks :) Did I overlook something? I could however, now that I think about it, remove the "(and newer)" as this doesn't make sense... :D

williamdes replied:
My point is dont force people to choose an option that are not into ;)

casperklein replied:
I think Georg just listed the platforms, that are **supported** by us.

georglauterbach replied:
As @casperklein said :D We _don't_ build images for these platforms. It is impossible to use such a platform unless you built it yourself.

williamdes replied:
But this is all the point ..
Please provide an "Other" option because people will choose one there are not into

polarathene replied:
> Maybe just add an "other" option

Soo... "unsupported" then? That'd be pretty clear ;)

williamdes replied:
> Soo... "unsupported" then? That'd be pretty clear ;)

Awesome, here is a great option

georglauterbach replied:
I see. I will add the "unsupported" option :)

georglauterbach replied:
Done. [6b6e11a00e29ff845119e20d6835e3896c89f0da]

williamdes commented on .github/ISSUE_TEMPLATE/bug_report_form.yml:
```diff
@@ -0,0 +1,209 @@
+      value: |
+        Thank you for participating in this project and reporting a bug. Docker Mail Server (DMS) is a community-driven project, and each contribution counts.
+
+        Please **fill out all the fields and checkboxes of this form** to make it easier for maintainers to understand the problem and to solve it. If this template is not adhered to, it will be marked with a label called `meta/no template - no support` and will not be worked on.
```
> If this template is not adhered to, it will be marked with a label called `meta/no template - no support` and will not be worked on.

Reading this make me feel pretty bad

And would probably make other people feel bad too about how the issue they are trying to report could be handled

Maybe we could add some kindness with a better phrasing ?

georglauterbach replied:
I am open for suggestions :)

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+    id: mailserver-version
+    attributes:
+      label: DMS version
+      description: With which version did yor encounter this bug? Paste your image tag here.
```
```suggestion
      description: On which version (image tag) did you encounter this bug?
```

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+    id: main-memory
+    attributes:
+      label: How much RAM is available to DMS explicitly?
+      description: Please provide the amount of memory that DMS can really use. This includes main memory (RAM), but also swap or zram. If you're using Kubernetes, provide the resource limit here if you set one.
```
Does this imply some k8s support?

```suggestion
      description: Please provide the amount of memory that DMS can really use. This includes physical memory (RAM), but also virtual memory (swap or zram). If you're using Kubernetes, provide the resource limit here if you set one.
```

williamdes commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+      value: |
+        Thank you for participating in this project and reporting a bug. Docker Mail Server (DMS) is a community-driven project, and each contribution counts.
+
+        Please **fill out all the fields and checkboxes of this form** to make it easier for maintainers to understand the problem and to solve it. If this template is not adhered to, it will be marked with a label called `meta/no template - no support` and will not be worked on.
```
```suggestion
        Please **fill out all the fields and checkboxes of this form** to make it easier for maintainers to understand the problem and to solve it.
        The maintainers and moderators are volunteers that need you to fill this template with accurate informations in order to help you in the best and quickest way.
        We will have to label your request with `meta/no template - no support` if your request is sloppy and provides no way to help you correctly.
```

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+  - type: dropdown
+    id: cpu-count
+    attributes:
+      label: How many CPU cores / threads are available?
+      description: How many CPU cores or threads can DMS use? If you're using Kubernetes, provide the resource limit here if you set one.
```
The only options are cores, do we need to account for threads separately or is cores sufficient?

Beyond k8s, you can set resource limits via Docker Compose AFAIK, possibly with `docker run` too, or with systemd cgroup driver on the target slice (daemon or container). "resource limit" is probably universally understood across all those by anyone that does it? (_although with cgroups you limit by cpu time I think, cores or threads would be unrelated pinning them to a process or assigning affinity, something like that, I'm not familiar with k8s limits_)

georglauterbach replied:
Done. [998cb4b4884b4efcce5930e2c24d7cc902e658aa]

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+      label: Is DMS running in a virtualized environment?
+      description: Choose the option that your setup is the most similar to. DMS is running on...
+      options:
+        - ... a physical host (with real, physical CPU cores)
```
aka Bare Metal? can dismiss, just suggesting due to the other two using title casing.

```suggestion
        - ... a Physical Host (with real, physical CPU cores)
```

georglauterbach replied:
I will instead make the other two lowercase 😆

georglauterbach replied:
Done. [998cb4b4884b4efcce5930e2c24d7cc902e658aa]

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+        - Linux
+        - macOS
+        - Windows
+        - Other
```
Seems more relevant :smile: 

```suggestion
        - Unsupported
```

georglauterbach replied:
macOS is not officially, and Windows by no means supported. "Unsupported" would therefore be a real superset of the former two :) But I have a solution in mind...

georglauterbach replied:
Done. [998cb4b4884b4efcce5930e2c24d7cc902e658aa]

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+        - Docker Compose
+        - Podman
+        - Kubernetes
+        - Other
```
```suggestion
        - Unsupported
```

georglauterbach replied:
Same as above, I think I have a solution

georglauterbach replied:
Done. [998cb4b4884b4efcce5930e2c24d7cc902e658aa]

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+    validations:
+      required: false
+  - type: input
+    id: uname-movar
+    attributes:
+      label: The output of `uname -movar`
+      description: Please paste the output of `uname -movar` here. If you're unable to do so (maybe because you're on Windows), tell us the reason for your inability here.
```
What happened here? `uname -movr` is pointless if doing `uname -a` as well.

If we just want the kernel version `uname -r` works well enough. If you're fine with the full output shown in your placeholder, keep it at `uname -a`.

georglauterbach replied:
I messed up. I just want `uname -a` to get the full view. It is also "easier", although I gotta admit this is a somewhat useless point. In the end, this is really here to double-check.

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+      label: Code of conduct
+      description: By submitting this issue, you agree to follow [our code of conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md).
+      options:
+        - label: I read this project's Code of Conduct and I agree
```
```suggestion
        - label: I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
```

georglauterbach replied:
Done. [998cb4b4884b4efcce5930e2c24d7cc902e658aa]

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+          required: true
+        - label: Just to make sure you're not just clicking checkboxes, please do not check this checkbox.
+          required: false
+        - label: I read the README and the documentation but could not find a solution
```
```suggestion
        - label: I have read the README and the documentation but could not find a solution
```

georglauterbach replied:
why the "have". Simple past is fine if you ask me :) It actually is, IIRC, better to use it here that present perfect because it implies a sense of having-already-done-it. But I guess this is about the fact that one cannot distinguish between simple present and simple past when reading "read". I will adjust this. (BTW this escalated quickly :D)

georglauterbach replied:
Done. [998cb4b4884b4efcce5930e2c24d7cc902e658aa]

williamdes commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,211 @@
+      value: |
+        Thank you for participating in this project and reporting a bug. Docker Mail Server (DMS) is a community-driven project, and each contribution counts.
+
+        Please **fill out all the fields and checkboxes of this form** to make it easier for maintainers to understand the problem and to solve it. The maintainers and moderators are volunteers that need you to fill this template with accurate informations in order to help you in the best and quickest way. We will have to label your request with `meta/no template - no support` if your request is  ...
```
```suggestion
        Please **fill out all the fields and checkboxes of this form** to make it easier for maintainers to understand the problem and to solve it.
        The maintainers and moderators are volunteers that need you to fill this template with accurate informations in order to help you in the best and quickest way.
        We will have to label your request with `meta/no template - no support` if your request is sloppy and provides no way to help you correctly.
```

georglauterbach replied:
I thought that this information belongs together and should therefore be "on the same line".

williamdes replied:
Oh okay, I am not sure where I can see a GUI of this config

georglauterbach replied:
Sadly, no one can. I'd like that too. We can always put lines in between, but I'd start off by trying to put all this in one line :D

polarathene replied:
"informations" -> "information"

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -113,9 +113,9 @@ body:
-        - macOS
-        - Windows
-        - Other
+        - macOS (not officially supported)
+        - Windows (unsupported)
+        - Other (unsupported)
```
Is there a difference between "not officially supported" and "unsupported"?

georglauterbach replied:
The idea was that some maintainers will be able to provide support here, while for `unsupported`, there are no maintainers to my knowledge that support these cases. I will add an explanation at the top.

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -136,9 +136,9 @@ body:
-        - Podman
-        - Kubernetes
-        - Other
+        - Podman (not officially supported)
+        - Kubernetes (not officially supported)
+        - Other (unsupported)
```
Same here. I assume you see value in having Podman and Kubernetes as separate options instead of a "Unsupported (Podman, Kubernetes, etc)"? I can understand why.

I know some maintainers have k8s experience; is that also the case with podman, or is it just common enough to provide some value here the report?

georglauterbach replied:
Same as above :)

georglauterbach replied:
Same as above :)

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -76,19 +76,19 @@ body:
     attributes:
       label: How much RAM is available to DMS explicitly?
-      description: Please provide the amount of memory that DMS can really use. This includes main memory (RAM), but also swap or zram. If you're using Kubernetes, provide the resource limit here if you set one.
+      description: Please provide the amount of memory that DMS can really use. This includes physical memory (RAM), but also virtual memory (like swap or zram).  Think about all possible resource restrictions on memory here.
```
Accidentally reviewing a commit, so I can't provide a suggestion here.

> Think about all possible resource restrictions on memory here.

Perhaps "Consider any resource limits you may have set that would affect this."

georglauterbach replied:
I like it.

polarathene commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -76,19 +76,19 @@ body:
-      label: How many CPU cores / threads are available?
-      description: How many CPU cores or threads can DMS use? If you're using Kubernetes, provide the resource limit here if you set one.
+      label: How many CPU cores are available?
+      description: How many CPU cores can DMS use? Think about all possible resource restrictions on CPU cores here.
```
> Think about all possible resource restrictions on CPU cores here.

Likewise "Consider any resource limits you may have set that would affect this."

georglauterbach replied:
Same as above :)

polarathene commented:
@williamdes @georglauterbach you had a discussion about previewing the form earlier.

If you visit the document on the branch, Github will detect it and render it accordingly :sunglasses: :

https://github.com/docker-mailserver/docker-mailserver/blob/issue-forms/.github/ISSUE_TEMPLATE/bug_report.yml

williamdes commented:
It's funny but nobody did ever consider some input for asking the user about I/O delay 🤔 
I have no real way to provide some data/measure about it, but this can really create terrible issues
Maybe a select between SSD/HDD/Other/Do not know could be useful but not required

NorseGaud commented:
> It's funny but nobody did ever consider some input for asking the user about I/O delay 🤔
> I have no real way to provide some data/measure about it, but this can really create terrible issues
> Maybe a select between SSD/HDD/Other/Do not know could be useful but not required

+1 this. I use AWS EFS for example and it's awfully slow.

polarathene commented:
> Maybe a select between SSD/HDD/Other/Do not know could be useful but not required

Another common disk source is probably network storage (eg via NFS, SMB, Ceph, etc). Not uncommon with VPS to have block storage attached separately if default disk capacity is too low for the VPS plan.

If necessary for more information, could consider requesting such below, but I think that'd just add friction that isn't required for most reports.

## Disk Info

You can get rough details about disks via something like `lsblk` but this is probably more info than required.

<details>
<summary>Disk and/or system info via inxi</summary>

[`inxi`](https://github.com/smxi/inxi) is a popular tool that can get a variety of system info, but may require user to install it. Should be available by each distro package manager. You can get disk info via `inxi --disk`:

```
Drives:    Local Storage: total: 1.02 TiB used: 499.63 GiB (47.9%) 
           ID-1: /dev/sda vendor: Crucial model: CT1000MX500SSD1 size: 931.51 GiB 
           ID-2: /dev/sdb vendor: Colorful model: SL300 120GB size: 111.79 GiB
```

That shows two SSD devices. Not sure if that includes network storage, `inxi --help` shows other commands for available disk space, partitions, if RAID is used etc. Extra detail can be requested up to 3 levels depending on options, 1 `x` per level, for disks you get link speed (eg 6 Gb/sec for SATA 3).

`inxi -p` for partitions and filesystem info:

```
Partition: ID-1: / size: 19.99 GiB used: 16.04 GiB (80.2%) fs: xfs dev: /dev/sdb5 
           ID-2: /boot/efi size: 96 MiB used: 24 MiB (25.0%) fs: vfat dev: /dev/sdb2 
           ID-3: /home size: 19.99 GiB used: 18.31 GiB (91.6%) fs: xfs dev: /dev/sdb6 
           ID-4: /run/media/polarathene/data size: 499.76 GiB used: 443.27 GiB (88.7%) fs: xfs dev: /dev/sda1 
           ID-5: /run/media/polarathene/projects size: 19.99 GiB used: 19.83 GiB (99.2%) fs: xfs 
           dev: /dev/sdb8 
           ID-6: swap-1 size: 4 GiB used: 2.15 GiB (53.7%) fs: swap dev: /dev/sdb7
```

Here is `inxi --basic` on one of my systems (tad outdated) that can provide a basic overview:

```
System:    Host: polarathene-linux Kernel: 5.4.89-1-MANJARO x86_64 bits: 64 Desktop: KDE Plasma 5.20.5 Distro: Manjaro Linux 
Machine:   Type: Desktop Mobo: ASRock model: H170 Performance/Hyper serial: <superuser required> UEFI: American Megatrends 
           v: P7.20A date: 01/23/2017 
CPU:       Info: Quad Core Intel Core i5-6500 [MCP] speed: 800 MHz min/max: 800/3600 MHz 
Graphics:  Device-1: NVIDIA GP104 [GeForce GTX 1070] driver: nvidia v: 460.32.03 
           Display: x11 server: X.Org 1.20.10 driver: loaded: nvidia resolution: 1: 1440x900~60Hz 2: 1920x1080~60Hz 
           OpenGL: renderer: GeForce GTX 1070/PCIe/SSE2 v: 4.6.0 NVIDIA 460.32.03 
Network:   Message: No Device data found. 
           Device-1: ASUSTek 802.11ac WLAN type: USB driver: mt76x2u 
Drives:    Local Storage: total: 1.02 TiB used: 499.63 GiB (47.9%) 
Info:      Processes: 560 Uptime: 102d 19h 44m Memory: 31.3 GiB used: 30.44 GiB (97.2%) Shell: Bash inxi: 3.2.02
```

The memory is a bit inaccurate I think as I have 8GB swap (4GB disk + 4GB zram), my system monitor says 27.9 / 31.3 GiB RAM and 4.7 GiB / 8 GiB SWAP (2.6G is zram, about 600M compressed).

</details>

---

## Measuring Disk Latency:

<details>
<summary>ioping</summary>

For latency measurements apparently [`ioping` should work](http://woshub.com/check-disk-performance-iops-latency-linux/). 

Output example from the link:

```
4 KiB <<< /tmp/ (ext4 /dev/md126p5): request=1 time=1.55 ms (warmup)
......................
4 KiB <<< /tmp/ (ext4 /dev/md126p5): request=19 time=176.3 us (fast)
4 KiB <<< /tmp/ (ext4 /dev/md126p5): request=20 time=356.9 us
--- /tmp/ (ext4 /dev/md126p5) ioping statistics ---
19 requests completed in 5.67 ms, 76 KiB read, 3.35 k iops, 13.1 MiB/s
generated 20 requests in 19.0 s, 80 KiB, 1 iops, 4.21 KiB/s
min/avg/max/mdev = 176.3 us / 298.7 us / 368.9 us / 45.7 us
```

However I'm not familiar with this tool and [another article has this warning](https://www.cyberciti.biz/faq/linux-freebsd-openbsd-macosx-find-disk-io-latency-with-ioping/):

> **Warning:** Use the ioping command with care. _Wrong options can destroy your data_. Read man page for more info.
> 
> Safe for directory target. Write I/O gives more reliable results for systems where non-cached reads are not supported or cached at some level. Might be *DANGEROUS* for file/device: it will shred your data. In this case should be repeated tree times (-WWW).

</details>

---

<details>
<summary>fio</summary>

`fio` also can be used to get a sense of latency measurements via `bs=4k iodepth=1` [according to Tencent Cloud](https://intl.cloud.tencent.com/document/product/362/6741).

</details>

georglauterbach commented:
@williamdes @polarathene @NorseGaud I will add a memory section for it.

casperklein commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,239 @@
+          required: true
+        - label: Just to make sure you're not just clicking checkboxes, please do not check this checkbox.
+          required: false
+        - label: I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) but could not find a solution
```
Maybe we should include "issues"?: "I've read the README, Documentation and (old) issues?

georglauterbach replied:
I can add this. We would then be consistent with the top of the form, where this was already said.

williamdes commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,239 @@
+      label: Miscellaneous first checks
+      description: Please read these carefully.
+      options:
+        - label: I checked that all ports are open and not blocked by my ISP / hoster.
```
```suggestion
        - label: I checked that all ports are open and not blocked by my ISP / hosting company.
```

polarathene replied:
company --> provider

wernerfred commented on target/scripts/start-mailserver.sh:
```diff
@@ -256,9 +256,8 @@ function _defunc
-_notify 'inf' 'ENVIRONMENT'
-[[ ${DMS_DEBUG} -eq 1 ]] && printenv
 
+[[ ${DMS_DEBUG} -eq 1 ]] && print_environment
```
What about not printing the env on startup but making this a callable script that can be executed if needed (e.g. in the form when it comes to the environment make a hint to run `docker exec -it mailserver print_environment` to get the env on runtime.

I see multiple benefits on this way as ENVs might getting changed after startup or you do not want to restart your server to fill an issue report (and may loose the actual flawed instance with logs)

georglauterbach replied:
I implemented this behavior. But we will now need to use the `.bashrc` file because we do not have the `${VARS}` array at hand. Done with 4255affabfa916bb4a38f490eb93e907ecc6b2e5.

georglauterbach replied:
Here you go:

https://github.com/docker-mailserver/docker-mailserver/blob/6336c0b256682527b880032301da4d3ba555e030/target/scripts/startup/setup-stack.sh#L70-L83

casperklein replied:
1. `docker exec mailserver cat /root/.bashrc` to get also the container ENVs
2. ENVs should not change during runtime, do they?
3. It doesn't hurt IMO, when `DMS_DEBUG` outputs the configured ENVs. PS: There might be people, letting DMS_DEBUG configured all the time and not only when there are issues 😉

wernerfred replied:
Ah thanks for clarification 👀 now i get it

georglauterbach replied:
Do I overlook something? ATM `$DMS_DEBUG=1` does print the environment variables and we have a script that does this so we're able to execute it at a later point in time too. Isn't this the best of both worlds?

casperklein replied:
Yep. The only thing is, do we need an additional script, when `docker exec mailserver cat /root/.bashrc` provides the same information? (PS: I am also fine, if we keep the script 😉)

georglauterbach replied:
I think this is the way to go, because `cat`ing the `.bashrc` does not format the chosen ENV variables as nice as my little addition does 😆

wernerfred reviewed (commented):
Not sure if deleting `.github/ISSUE_TEMPLATE/bug_report.md` should be done right now. Maybe let bot method live next to each other for a bit of time to evaluate how it is used and if some options are missing before removing the legacy method. (Although keep in mind that issue forms are beta and subject to change - maybe not rely heavily on it).

georglauterbach commented:
> Not sure if deleting `.github/ISSUE_TEMPLATE/bug_report.md` should be done right now. Maybe let bot method live next to each other for a bit of time to evaluate how it is used and if some options are missing before removing the legacy method. (Although keep in mind that issue forms are beta and subject to change - maybe not rely heavily on it).

I would like to let the other templates (PR / Other) live next to the bug report form, because I don't see a benefit in having two bug report form, when we're ultimately interested in feedback on the issue form. By only having the form for issues we will receive more feedback.

georglauterbach commented:
Why in hell do test fail _after rebasing onto master_??? This is a docs update, mostly, and we have no test that scans ENV variable printing... 😆 Does someone know?

polarathene commented:
> Why in hell do test fail _after rebasing onto master_???

It's the same recurring failing test. It's not a particularly reliable one, perhaps we should disable it until someone has time to fix it? I mean if it isn't deterministic, it's not much of a valid test is it?

`wait_for_empty_mail_queue_in_container fails when timeout reached`:

```
not ok 209 wait_for_empty_mail_queue_in_container fails when timeout reached
# (from function `repeat_until_success_or_timeout' in file test/test_helper/common.bash, line 27,
#  from function `repeat_in_container_until_success_or_timeout' in file test/test_helper/common.bash, line 70,
#  from function `wait_for_empty_mail_queue_in_container' in file test/test_helper/common.bash, line 192,
#  in test file test/test_helper.bats, line 225)
#   `! TEST_TIMEOUT_IN_SECONDS=0 wait_for_empty_mail_queue_in_container "${CONTAINER_NAME}"' failed
# 220 mail.my-domain.com ESMTP
# 250 mail.my-domain.com
# 250 2.1.0 Ok
# 250 2.1.5 Ok
# 354 End data with <CR><LF>.<CR><LF>
# 250 2.0.0 Ok: queued as 3685E2F9F4B
# 221 2.0.0 Bye
# 55d07f97cc479d22a0e1e6fff8b13c2ba305db0fb140522788fe2dd2f7b338cb
```

and one after it `wait_for_empty_mail_queue_in_container succeeds within timeout`:

```
not ok 210 wait_for_empty_mail_queue_in_container succeeds within timeout
# (in test file test/test_helper.bats, line 250)
#   `[[ ${SECONDS} -gt 0 ]]' failed
# 220 mail.my-domain.com ESMTP
# 250 mail.my-domain.com
# 250 2.1.0 Ok
# 250 2.1.5 Ok
# 354 End data with <CR><LF>.<CR><LF>
# 250 2.0.0 Ok: queued as 9C7BE2F9F4A
# 221 2.0.0 Bye
# 56af3bbce5a658026b72553e87d349fff017b3a5a2dd26a8965bd265ed4c34fd
```

I'm not familiar with the test or how to debug that to make it deterministic.

NorseGaud commented:
Hi guys, pull master branch of bats and run it with "--verbose-run --trace". It may help show what exactly is being run and what is not working to reproduce it locally.

I wouldn't recommend disabling a test if it's intent is to prove that something works as it should.

polarathene commented:
> I wouldn't recommend disabling a test if it's intent is to prove that something works as it should.

IIRC it's been a test that has failed throughout the year with no code changes that affect it.

If the test is not deterministic, it's not reliable to rely on. A pass means little at that point.

I didn't advocate for it's removal, just disabled since it's failure contributes no value either where the solution is to just run the test suite again until it passes despite no changes made.

georglauterbach commented:
I restarted the jobs one more time. If this run fails again, I agree with @polarathene at this point that we should disable these two tests and investigate...

georglauterbach commented:
Tests pass. Leave your reviews again :D

polarathene commented:
No rush to merge this right? I have dismissed @wernerfred stale review, but I think we can leave this open over the weekend in case he still wants to chime in. If no other activity merge from Monday?

wernerfred reviewed (approved):
Sorry for not responding for a week now but i needed some time offline 🏞️ 

I looked at the rendering of the form and feel like that the `<=` syntax is not looking nice imho. I proposed a written variant but feel free to change the wording.

Thanks @polarathene for thinking of me and giving me time to throw in my two cents once again. 🚀 

Feel free to merge if you disagree with my changes.

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,239 @@
+      options:
+        - <= 1GB (very low)
+        - <= 2GB
+        - <= 4GB
```
Done. [a5b6018ba8ecfb4966a301681001b19af0209bca]

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,239 @@
+      options:
+        - label: I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
+          required: true
+        - label: Just to make sure you're not just clicking checkboxes, please do not check this checkbox.
```
Done. [a5b6018ba8ecfb4966a301681001b19af0209bca]

georglauterbach commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,239 @@
+      options:
+        - <= 1 Core
+        - <= 2 Cores
+        - <= 4 Cores
```
Done. [a5b6018ba8ecfb4966a301681001b19af0209bca]

georglauterbach commented:
Aright. I incorporated @wernerfred's feedback and added a feedback text area field to the forms for users to tell us their impressions too. I hope that all is well now :) Please leave your reviews (for the last time I hope) again :D

wernerfred reviewed (approved):
Nice idea with the feedback section 👍🏻
